### PR TITLE
Add `PreventUsingIncubatingMethods` error-prone check (conjure)

### PR DESCRIPTION
**Pull request by blacksmithgu**
An explicit check which warns against using incubating APIs; you can disable it on a per-project basis, or on a per-usage basis using @SuppressWarnings.

## Before this PR
Incubating APIs are a stability liability, and should generally only be used explicitly (i.e., opt-in instead of opt-out). This check prevents their usage without an explicit annotation or opt-out.

## After this PR
Adds the `PreventUsingIncubatingMethods` error-prone check, which blocks methods annotated with Conjure's `@Incubating` annotation.

## Possible downsides?
Users upgrading `gradle-baseline` may get errors related to incubating methods until they either disable the check or tag their incubating API usages.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/gradle-baseline](https://github.com/palantir/gradle-baseline), @blacksmithgu! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by blacksmithgu** [comment]
cc @carterkozak

File: README.md [hunk 1]
```diff
@@ -198,6 +198,7 @@ Safe Logging can be found at [github.com/palantir/safe-logging](https://github.c
 - `UnnecessarilyQualified`: Types should not be qualified if they are also imported.
 - `DeprecatedGuavaObjects`: `com.google.common.base.Objects` has been obviated by `java.util.Objects`.
 - `JavaTimeSystemDefaultTimeZone`: Avoid using the system default time zone.
+- `PreventUsingIncubatingMethods`: Prevents calling Conjure incubating APIs unless you explicitly opt-out of the check on a per-use or per-project basis.
```
Comment (carterkozak) [inline README.md]: Thoughts on renaming this to something like `IncubatingMethod`? We haven't been great about naming, but most error-prone checks are named to describe the bad thing they detect.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import com.sun.tools.javac.code.Symbol;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "PreventUsingIncubatingMethods",
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "You should avoid using incubating methods where possible, since they have very weak stability"
+                + " guarantees. You can explicitly disable this check on a case-by-case basis using"
+                + " @SuppressWarnings(\"PreventUsingIncubatingMethods\"), or disable it globally for new projects.")
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]:
```suggestion
                + " @SuppressWarnings(\"PreventUsingIncubatingMethods\").")
```

Let's leave the global piece implied, we don't want to give folks too many ideas!

File: README.md (hunk 1 above)
Comment (blacksmithgu) [reply README.md]: That's fine. I don't think we have code using the suppression yet, so no breakages.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    public Description matchMemberReference(MemberReferenceTree tree, VisitorState state) {
+        return checkTree(tree, ASTHelpers.getSymbol(tree), state);
+    }
+
+    private Description checkTree(Tree tree, Symbol.MethodSymbol method, VisitorState state) {
+        if (!ASTHelpers.hasAnnotation(method, INCUBATING, state)) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree)
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]: No need to set a message, the summary from the annotation above is sufficient
```suggestion
        return describeMatch(tree);
```

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java (hunk 2 above)
Comment (blacksmithgu) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]: Alright.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java (hunk 3 above)
Comment (blacksmithgu) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]: Ack, didn't know that method existed.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return checkTree(tree, ASTHelpers.getSymbol(tree), state);
+    }
+
+    @Override
+    public Description matchMemberReference(MemberReferenceTree tree, VisitorState state) {
+        return checkTree(tree, ASTHelpers.getSymbol(tree), state);
+    }
+
+    private Description checkTree(Tree tree, Symbol.MethodSymbol method, VisitorState state) {
+        if (!ASTHelpers.hasAnnotation(method, INCUBATING, state)) {
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]: this check can be extracted to a static reusable matcher:
```java
Matchers.hasAnnotation("com.palantir.conjure.java.lib.internal.Incubating");
```

We should probably not fail builds when folks test their endpoints, so we can avoid reporting a match if `TestCheckUtils.isTestCode(state)` is true (note that this method is relatively expensive, so we should only check it after we've found an incubating match).

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/IncubatingMethodTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+package com.palantir.baseline.errorprone;
+
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.Test;
+
+public class IncubatingMethodTest {
+
+    // We define our own simple conjure incubating annotation at the right path:
+    private static final String[] ANNOTATION_DEFINITION =
+            new String[] {"package com.palantir.conjure.java.lib.internal;", "public @interface Incubating {}"};
```
Comment (carterkozak) [inline baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/IncubatingMethodTest.java]: Might be simpler to add a `testCompile` dependency on `conjure-lib` so we don't have to define this in tests, and we know we're using the right fqcn

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java (hunk 4 above)
Comment (blacksmithgu) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreventUsingIncubatingMethods.java]: That is a good idea.

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/IncubatingMethodTest.java (hunk 5 above)
Comment (blacksmithgu) [reply baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/IncubatingMethodTest.java]: Sure, that makes sense. Had to remember how to use `version.props` correctly...

**Review by carterkozak — APPROVED** [review]
Lgtm, thanks!

**Pull request merged**